var/spack/repos/builtin/packages/llvm-doe/package.py

```diff
@@ -0,0 +1,525 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import os.path
+import re
+import sys
+
+import llnl.util.tty as tty
+import spack.util.executable
+
+
+class LlvmDoe(CMakePackage, CudaPackage):
+    """This package provides a collection of the experimental LLVM projects done
+    by the US DOE research and development teams.
+    """
+
+    homepage = "https://github.com/llvm-doe-org"
+    url = "https://github.com/llvm-doe-org/llvm-project/archive/llvmorg-10.0.0.zip"
+    git = "https://github.com/llvm-doe-org/llvm-project"
+    maintainers = ['shintaro-iwasaki']
+
+    family = "compiler"  # Used by lmod
+
+    version('main', branch='doe', preferred=True)
+    version('upstream', branch='llvm.org/main')
+    version('bolt', branch='bolt/main')
+    version('clacc', branch='clacc/master')
+    version('pragma-clang-loop', branch='sollve/pragma-clang-loop')
+    version('pragma-omp-tile', branch='sollve/pragma-omp-tile')
+
+    # NOTE: The debug version of LLVM is an order of magnitude larger than
+    # the release version, and may take up 20-30 GB of space. If you want
+    # to save space, build with `build_type=Release`.
+
+    variant(
+        "clang",
+        default=True,
+        description="Build the LLVM C/C++/Objective-C compiler frontend",
+    )
```

Comment:
In general most of the code below seems duplicated from the `llvm` package. Would it be possible to coordinate with `llvm` maintainers and reduce code duplication to a minimum here? I think in this case it's particularly relevant, since the package recipe for `llvm` is fairly long and changes frequently.

Reply:
I understand copying all is controversial, but we would like to maintain this package independently from the official LLVM package.

We'd like to do so because the upstream LLVM Spack package development is faster than that of this package.  If this package becomes a subclass of `Llvm`, for example, this package must be updated whenever the upstream LLVM Spack package is changed and breaks the compatibility.  As far as I tried previously, making a subclass of LLVM can surely reduce the duplication but it is fragile since this package class needs to depend on the internal variables and functions of LLVM. The code was not very clean since the LLVM package is not designed as a base class of another package (and I don't think the `llvm` maintainers want to redesign it).

A similar discussion happened in https://github.com/spack/spack/pull/12607, and at that time, the Spack developers allowed us to duplicate the code.

Reply:
> I understand copying all is controversial, but we would like to maintain this package independently from the official LLVM package.

In that case, do you need all the variants and choices from upstream or can you simplify this recipe?

Reply:
We would like to keep all the variants so that one can install this as (at least at present) a replacement of `llvm` though I understand some configurations are rarely changed by users.